colexecbase: refactor cast templating to be type-specific

yuzefovich opened the pull request:
See individual commits for details.

Informs: #48135.

yuzefovich committed:
colexecbase: clean up identity cast operators

This commit introduces a custom `castIdentityOp` operator that handles
casts when "from" and "to" types are identical. The job of such operator
is to simply copy the contents of one vector into another, without
performing the deselection step. This allows us to remove some pretty
redundant code of old identity cast operators (and the new code will
also be faster).

Not performing the deselection is justified by the following:
1. to be in line with other cast operators
2. AND/OR projection operators cannot handle when a different batch is
   returned than the one they fed into the projection chain (which might
   contain casts)
3. performing the deselection would require copying over all vectors, not
   just the output one.
This operator should be planned rarely enough (if ever) to not be very
important.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66459)
<!-- Reviewable:end -->

yuzefovich committed:
colexecbase: refactor cast templating to be type-specific

This commit refactors our templating of cast overloads from "canonical
type family + width" specific to "actual type" specific. This will allow
us support more casts than before (previous mapping was too limited).
This commit attempts to make the minimal changes to the generated code
for ease of reviewing. The follow-up commit will clean some things up.
Note that this commit required introducing the explicit support for
casts from dates (previously we were getting that implicitly since dates
have the same representation as INT8).

The support of casts from datum-backed types is removed in this commit
but will be added back in the follow-ups.

This refactor required that we set up the 4-leveled hierarchy (similar
to what we have for twoArgsOverloads) for the supported casts between
natively supported types.

Release note: None

yuzefovich committed:
execgen: remove leftover UNSAFEGET stuff

Some of `UNSAFEGET` things in the template generators are a left-over of
49780f3fd7d2e3dfb012575ada44e2c9ad0ec6b1 and can be safely removed since
they are no longer used/needed.

Release note: None

yuzefovich committed:
colexecbase: bring back the support for some casts from datums

The previous commit removed the support of casts between datum-backed
types and from datums to bools (in order to simplify the refactor). This
commit adds those casts back.

Casts from datum-backed types to natively supported ones required
setting up the 2-leveled hierarchy (only on the "to" side) whereas the
casts between datum-backed types can be supported by a single operator.
This commit utilizes the same template for the body of the operator for
all kinds of cast operators.

Release note: None

yuzefovich committed:
colexecbase: clean up cast operators a bit

This commit extracts a common cast base struct (responsible for
implementing `Resetter` interface) which contains the initialization
fields. It also addresses some stylistic TODOs.

Additionally, this commit restructures the "resolution" of which cast
operator to use (to separate out datum-backed vs native types on both
sides of the cast) as well as updates the template for the casts to
native types to set the length on the batch (which is required for
bytes-like vector - although we haven't added support for such cases
yet).

Release note: None

yuzefovich commented:
I think this is RFAL.

I tried to be careful with splitting up the changes so that they could be reviewed more easily. I think in many cases it is worth looking at the actual generated code in `cast.eg.go` file.

DrewKimball commented on pkg/sql/colexec/colexecbase/cast_tmpl.go:
```diff
@@ -85,19 +83,28 @@ func GetCastOperator(
+	if toType.Identical(fromType) {
+		// We have an identity cast, so we use a custom identity cast operator.
+		return &castIdentityOp{
+			OneInputInitCloserHelper: colexecop.MakeOneInputInitCloserHelper(input),
```
[nit] Since the last commit is for cleanup anyway, it might be nice to pull the common fields out into an embedded struct. What do you think?

DrewKimball reviewed (commented):
My reading is that previously, we treated `Date` columns the same as `Int` columns because they have the same canonical type family, but now that we switch on the actual types, we don't get this behavior. (I don't see `Date` anywhere in the type switches). Maybe that wasn't desired behavior in the first place, I'm not sure. Am I missing something here?

yuzefovich reviewed (commented):
Nice catch about the dates - I think it was never intentional that casts from dates were supported in the first place, but I checked the code, and it is basically correct (modulo known shortcoming of incorrectly handling dates outside of the valid range), so I adjusted the refactor commit to bring back the support of casts from dates.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-:-MezP_tu8afqtZiCg4jL:b-uyot59)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 89 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-McC-uEM8x8ct88DChz6-r9-89:-MezOKRQCwHOBk_gZXOR:b4ptnmn) ([raw file](https://github.com/cockroachdb/cockroach/blob/00183c0089f6518f0a43ff30ada40e77c5457efe/pkg/sql/colexec/colexecbase/cast_tmpl.go#L89)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] Since the last commit is for cleanup anyway, it might be nice to pull the common fields out into an embedded struct. What do you think?
</blockquote></details>

Good point. I actually had such a commit already brewing in #67733, but it makes sense to squash them together. Done.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66459)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @yuzefovich)

___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 109 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-MezrCTB-HkS8dP3M1yA:-MezrCTB-HkS8dP3M1yB:b8i6e52) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea2da143f125ae412333d238fc13e0c066140ee7/pkg/sql/colexec/colexecbase/cast_tmpl.go#L109)):*
> ```Go
> 			// {{end}}
> 		}
> 	} else {
> ```

[nit] I think you could get rid of the `else` block and even the `!isToDatum` check and just put this switch statement after the `isFromDatum` block - might look a bit simpler. Up to you, though. Barring that, it could be `else if !istoDatum` instead of the nested `if`.
___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 162 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-MezrEuj-oJt3zxLzRTC:-MezrEuj-oJt3zxLzRTD:b7skc4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea2da143f125ae412333d238fc13e0c066140ee7/pkg/sql/colexec/colexecbase/cast_tmpl.go#L162)):*
> ```Go
> 			// {{end}}
> 		}
> 	} else {
> ```

[nit] similar case as above
___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 357 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-MezruB63J6G4wD8hxhq:-MezruB63J6G4wD8hxhr:bmrvoij) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea2da143f125ae412333d238fc13e0c066140ee7/pkg/sql/colexec/colexecbase/cast_tmpl.go#L357)):*
> ```Go
> 			// invariant of flat bytes).
> 			// */}}
> 			batch.SetLength(n)
> ```

Good catch here.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-:-Mf-93SX5t65donU3oiZ:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 109 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-MezrCTB-HkS8dP3M1yA:-Mf-8ADm917lXqn8OIHG:bn3v2cv) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea2da143f125ae412333d238fc13e0c066140ee7/pkg/sql/colexec/colexecbase/cast_tmpl.go#L109)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] I think you could get rid of the `else` block and even the `!isToDatum` check and just put this switch statement after the `isFromDatum` block - might look a bit simpler. Up to you, though. Barring that, it could be `else if !istoDatum` instead of the nested `if`.
</blockquote></details>

I've intentionally kept this somewhat redundant structure so that it was clear which kinds of casts are currently supported. What I mean is that we have 4 cases:
1. native to native
2. native to datum
3. datum to native
4. datum to datum

but at the moment we partially support 1 and 3 while fully supporting 4. In the follow-up PR we'll complete the support of 3 and add the support for 2 (thus only leaving 1 partially implemented).

So I'll keep the structure as is since in the follow-up PR the redundant structure will be filled in.
___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 162 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-MezrEuj-oJt3zxLzRTC:-Mf-92NH0lar1_dXFI9x:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea2da143f125ae412333d238fc13e0c066140ee7/pkg/sql/colexec/colexecbase/cast_tmpl.go#L162)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

[nit] similar case as above
</blockquote></details>

Done.
___
*[pkg/sql/colexec/colexecbase/cast_tmpl.go, line 357 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/66459#-MezruB63J6G4wD8hxhq:-Mf-84EvDYKaZA6eRtY4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea2da143f125ae412333d238fc13e0c066140ee7/pkg/sql/colexec/colexecbase/cast_tmpl.go#L357)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Good catch here.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
